core/services/relay/evm/bindings.go

```diff
@@ -5,57 +5,65 @@ import (
 	"fmt"
 
 	commontypes "github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
 )
 
-// key is contract name
-type contractBindings map[string]readBindings
+// bindings manage all contract bindings, key is contract name.
+type bindings map[string]*contractBinding
 
-// key is read name
-type readBindings map[string]readBinding
-
-func (b contractBindings) GetReadBinding(contractName, readName string) (readBinding, error) {
-	rb, rbExists := b[contractName]
-	if !rbExists {
+func (b bindings) GetReadBinding(contractName, readName string) (readBinding, error) {
+	// GetReadBindings should only be called after Chain Reader init.
+	cb, cbExists := b[contractName]
+	if !cbExists {
 		return nil, fmt.Errorf("%w: no contract named %s", commontypes.ErrInvalidType, contractName)
 	}
 
-	reader, readerExists := rb[readName]
-	if !readerExists {
+	rb, rbExists := cb.readBindings[readName]
+	if !rbExists {
 		return nil, fmt.Errorf("%w: no readName named %s in contract %s", commontypes.ErrInvalidType, readName, contractName)
 	}
-	return reader, nil
+	return rb, nil
 }
 
-func (b contractBindings) AddReadBinding(contractName, readName string, reader readBinding) {
-	rbs, rbsExists := b[contractName]
-	if !rbsExists {
-		rbs = readBindings{}
-		b[contractName] = rbs
+func (b bindings) AddReadBinding(contractName, readName string, rb readBinding) {
+	// Adding read bindings outside of Chain Reader init is not thread safe.
```

Comment:
nit/ mv this to be a comment on the func so that it is displayed in godoc & ide help